flocker/provision/_ssh/_conch.py

```diff
@@ -0,0 +1,171 @@
+from pipes import quote as shell_quote
+from twisted.python import log
+
+from characteristic import attributes
+
+from effect import (
+    sync_performer, TypeDispatcher, ComposedDispatcher, Effect,
+    )
+from effect.twisted import (
+    perform, deferred_performer)
+from twisted.conch.endpoints import (
+    SSHCommandClientEndpoint, _NewConnectionHelper, _ReadFile, ConsoleUI)
+
+from twisted.conch.ssh.keys import Key
+from twisted.internet import reactor
+from twisted.internet.defer import Deferred, inlineCallbacks
+from twisted.internet.endpoints import UNIXClientEndpoint, connectProtocol
+from twisted.internet.error import ConnectionDone
+from twisted.protocols.basic import LineOnlyReceiver
+from twisted.python.filepath import FilePath
+import os
+
+from flocker.testtools import loop_until
+
+from ._model import Run, Sudo, Put, Comment, RunRemotely
+
+
+@attributes([
+    "deferred",
+    "address",
+    "username",
+])
+class CommandProtocol(LineOnlyReceiver, object):
+    """
+    Protocol that logs the lines of a remote command.
+
+    :ivar Deferred deferred: Deferred to fire when the command finishes
+        If the command finished succesfuly, will fire with ``None``.
+        Otherwise, errbacks with the reason.
+    :ivar bytes username: For logging.
+    :ivar bytes address: For logging.
+    """
+    delimiter = b'\n'
+
+    def connectionMade(self):
+        self.transport.disconnecting = False
+
+    def connectionLost(self, reason):
+        if reason.check(ConnectionDone):
+            self.deferred.callback(None)
+        else:
+            self.deferred.errback(reason)
+
+    def lineReceived(self, line):
+        log.msg(format="%(line)s",
+                system="SSH[%s@%s]" % (self.username, self.address),
+                username=self.username, address=self.address, line=line)
+
+
+@sync_performer
+def perform_sudo(dispatcher, intent):
+    """
+    See :py:class:`Sudo`.
+    """
+    return Effect(Run(command='sudo ' + intent.command))
```

Comment:
How does this `sync_performer` interact with the `deferred_performer` that it wraps (in this dispatcher)? 

Does `sync_performer` wait for the deferred to fire before returning?